[perf only] Test function specialization

nikic opened the pull request:
r? @ghost

nikic committed:
update submodule

nikic commented:
@bors try @rust-timer queue

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit b4eef9bb78ce1a5d5ff713eb931b5fc474ae9ddc with merge 42f15a888b84383660493ba571b9d8262d73681a...
<!-- homu: {"type":"TryBuildStarted","head_sha":"b4eef9bb78ce1a5d5ff713eb931b5fc474ae9ddc","merge_sha":"42f15a888b84383660493ba571b9d8262d73681a"} -->

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3741274043/jobs/6351184740)
<!-- homu: {"type":"TryBuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/3741274043/jobs/6351184740","builder_name":"checks-actions"} -->

nikic committed:
Don't explicitly set C++ standard for lld

LLVM does this itself since 606cb8548a1b7763e0c8489c5efe66803a7ede72,
and 14 is no longer the correct standard when building lld 16,
causing build failures.

nikic committed:
Use LLVM_CMAKE_DIR for lld build

LLVM_CONFIG_PATH is no longer supported.

nikic commented:
LLD build still fails -- it looks like the `llvm-driver-template.cpp.in` export doesn't work for some reason.

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 2c8117139168b1cc8a9048b703cab0634d0d9e1c with merge 70212127e0974013f7e031861d421356e85b0e27...
<!-- homu: {"type":"TryBuildStarted","head_sha":"2c8117139168b1cc8a9048b703cab0634d0d9e1c","merge_sha":"70212127e0974013f7e031861d421356e85b0e27"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3748559099/jobs/6368182028)
Build commit: 70212127e0974013f7e031861d421356e85b0e27 (`70212127e0974013f7e031861d421356e85b0e27`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3748559099/jobs/6368182028"},"merge_sha":"70212127e0974013f7e031861d421356e85b0e27"} -->

rust-timer commented:
Queued 70212127e0974013f7e031861d421356e85b0e27 with parent 1d12c3cea30b8ba4a09650a9e9c46fe9fbe25f0b, future [comparison URL](https://perf.rust-lang.org/compare.html?start=1d12c3cea30b8ba4a09650a9e9c46fe9fbe25f0b&end=70212127e0974013f7e031861d421356e85b0e27).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (70212127e0974013f7e031861d421356e85b0e27): [comparison URL](https://perf.rust-lang.org/compare.html?start=1d12c3cea30b8ba4a09650a9e9c46fe9fbe25f0b&end=70212127e0974013f7e031861d421356e85b0e27&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=1d12c3cea30b8ba4a09650a9e9c46fe9fbe25f0b&end=70212127e0974013f7e031861d421356e85b0e27&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.9%  | [0.2%, 2.8%]   | 67    |
| Regressions ❌ <br /> (secondary)  | 0.9%  | [0.2%, 2.7%]   | 27    |
| Improvements ✅ <br /> (primary)   | -0.8% | [-4.9%, -0.2%] | 61    |
| Improvements ✅ <br /> (secondary) | -1.0% | [-5.6%, -0.3%] | 78    |
| All ❌✅ (primary)                 | 0.1%  | [-4.9%, 2.8%]  | 128   |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=1d12c3cea30b8ba4a09650a9e9c46fe9fbe25f0b&end=70212127e0974013f7e031861d421356e85b0e27&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.3%  | [0.8%, 3.6%]   | 10    |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.0% | [-3.9%, -0.1%] | 2     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 1.6%  | [-3.9%, 3.6%]  | 12    |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=1d12c3cea30b8ba4a09650a9e9c46fe9fbe25f0b&end=70212127e0974013f7e031861d421356e85b0e27&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 2.0%  | [1.1%, 2.7%]   | 11    |
| Regressions ❌ <br /> (secondary)  | 2.2%  | [2.1%, 2.4%]   | 2     |
| Improvements ✅ <br /> (primary)   | -2.0% | [-4.0%, -1.2%] | 98    |
| Improvements ✅ <br /> (secondary) | -2.9% | [-6.9%, -0.8%] | 51    |
| All ❌✅ (primary)                 | -1.6% | [-4.0%, 2.7%]  | 109   |
</details>

nikic committed:
baseline

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 228430a0ef3125c8056bddb91227f1433c88571c with merge 413335368bf860e701f6365b0f2f0fa4d7f85b9a...
<!-- homu: {"type":"TryBuildStarted","head_sha":"228430a0ef3125c8056bddb91227f1433c88571c","merge_sha":"413335368bf860e701f6365b0f2f0fa4d7f85b9a"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3750175054/jobs/6372254357)
Build commit: 413335368bf860e701f6365b0f2f0fa4d7f85b9a (`413335368bf860e701f6365b0f2f0fa4d7f85b9a`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3750175054/jobs/6372254357"},"merge_sha":"413335368bf860e701f6365b0f2f0fa4d7f85b9a"} -->

rust-timer commented:
Queued 413335368bf860e701f6365b0f2f0fa4d7f85b9a with parent 1d12c3cea30b8ba4a09650a9e9c46fe9fbe25f0b, future [comparison URL](https://perf.rust-lang.org/compare.html?start=1d12c3cea30b8ba4a09650a9e9c46fe9fbe25f0b&end=413335368bf860e701f6365b0f2f0fa4d7f85b9a).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (413335368bf860e701f6365b0f2f0fa4d7f85b9a): [comparison URL](https://perf.rust-lang.org/compare.html?start=1d12c3cea30b8ba4a09650a9e9c46fe9fbe25f0b&end=413335368bf860e701f6365b0f2f0fa4d7f85b9a&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=1d12c3cea30b8ba4a09650a9e9c46fe9fbe25f0b&end=413335368bf860e701f6365b0f2f0fa4d7f85b9a&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.9%  | [0.2%, 2.6%]   | 71    |
| Regressions ❌ <br /> (secondary)  | 0.9%  | [0.2%, 2.8%]   | 27    |
| Improvements ✅ <br /> (primary)   | -0.9% | [-5.2%, -0.2%] | 64    |
| Improvements ✅ <br /> (secondary) | -1.0% | [-5.6%, -0.2%] | 82    |
| All ❌✅ (primary)                 | 0.0%  | [-5.2%, 2.6%]  | 135   |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=1d12c3cea30b8ba4a09650a9e9c46fe9fbe25f0b&end=413335368bf860e701f6365b0f2f0fa4d7f85b9a&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.8%  | [1.4%, 2.1%]   | 5     |
| Regressions ❌ <br /> (secondary)  | 1.5%  | [1.5%, 1.5%]   | 1     |
| Improvements ✅ <br /> (primary)   | -1.8% | [-3.5%, -0.1%] | 2     |
| Improvements ✅ <br /> (secondary) | -1.6% | [-1.6%, -1.6%] | 2     |
| All ❌✅ (primary)                 | 0.7%  | [-3.5%, 2.1%]  | 7     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=1d12c3cea30b8ba4a09650a9e9c46fe9fbe25f0b&end=413335368bf860e701f6365b0f2f0fa4d7f85b9a&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 1.8%  | [1.2%, 2.5%]   | 14    |
| Regressions ❌ <br /> (secondary)  | 2.3%  | [2.2%, 2.4%]   | 2     |
| Improvements ✅ <br /> (primary)   | -2.0% | [-4.1%, -1.2%] | 91    |
| Improvements ✅ <br /> (secondary) | -3.1% | [-7.7%, -1.0%] | 54    |
| All ❌✅ (primary)                 | -1.5% | [-4.1%, 2.5%]  | 105   |
</details>

nikic commented:
Diff func spec against LLVM 16 baseline: https://perf.rust-lang.org/compare.html?start=413335368bf860e701f6365b0f2f0fa4d7f85b9a&end=70212127e0974013f7e031861d421356e85b0e27&stat=instructions%3Au
